### ISSUE TITLE
"No network" error when attempting to connect to a local, non-public synapse server via HTTP

### ISSUE THREAD

**Issue by Burchmeister** [vector-im/element-android#6721]
### Steps to reproduce

Basically, having the same issue noted in this error report below from a year ago. I would really like to have a way to work around this error. Being local only homeserver, I would like to avoid having to make SSL certs for a single user at this time. I have been able to link all other devices that are on iOS without issue. I receive the same "No Network" error, but no further information. The local network and VPN work flawlessly otherwise. 

My server address is: 'http://synapse.lan:8008'

I immediately receive a No Network error on Anroid only, exactly the same issue noted in the report below. 

Thank you in advance. 

https://github.com/vector-im/element-android/issues/3280#issue-876271603

#### What did you expect to happen?

_Similar to iOS, Windows and Linux applications, connection to server established, and moving to a login screen._

#### What happened instead?
I receive a "No network" error with no further error codes or reasoning. Rest of network connections aside from element-android work flawlessly. 

### Your phone model

Samsung S9

### Operating system version

Android 12

### Application version and app store

element-android 1.4.27

### Homeserver

Synapse, local-only, non-federated, non-public facing via http

### Will you send logs?

No

### Are you willing to provide a PR?

No

**Comment by ouchadam** [comment]
thanks for raising, out of interest does the login work if you use the ip address directly?

when developing we often use local synapse instances eg `http://10.0.2.2:8080` for the android emulator

**Comment by Burchmeister** [comment]
Thank you for your reply, I apologize for the delay. To answer your question, yes, I do have the same issue when using a direct IP connection on a LAN, and even via a wireguard VPN (as an experiement) to the same server. The issue is not able to be replicated on Windows, Linux, or iOS with the same login information.

**Comment by Burchmeister** [comment]
I believe that this will resolve the issue:

Resolution

If your application must support loading plain http:// URLs, you can opt into enabling cleartext traffic by using an AndroidManifest.xml file that includes an android:usesCleartextTraffic="true" application attribute. You can do that with the following steps.

    Download this example [AndroidManifest.xml](https://support.vuplex.com/article-assets/20200809-how-to-enable-cleartext-traffic-on-android/AndroidManifest.xml) file that includes the needed android:usesCleartextTraffic="true" attribute.

    Place the file in your project at Assets/Plugins/Android/AndroidManifest.xml. This location must be exact or else Unity will not detect its presence.

    Rebuild your application.

Source: https://support.vuplex.com/articles/how-to-enable-cleartext-traffic-on-android

**Comment by ouchadam** [comment]
we have a network config set up to allow for `clearTextTraffic` on a set of domains, such as`.lan` https://github.com/vector-im/element-android/blob/develop/vector/src/main/res/xml/network_security_config.xml#L8 although it's possible this configuration needs adjusting

another server configuration to check is that your wellknown http://synapse.lan:8008/.well-known/matrix/client is also returning a `base_url` on the `synapse.lan` domain

**Comment by Burchmeister** [comment]
> we have a network config set up to allow for `clearTextTraffic` on a set of domains, such as`.lan` https://github.com/vector-im/element-android/blob/develop/vector/src/main/res/xml/network_security_config.xml#L8 although it's possible this configuration needs adjusting
> 
> another server configuration to check is that your wellknown http://synapse.lan:8008/.well-known/matrix/client is also returning a `base_url` on the `synapse.lan` domain

I checked the well-known, and did not get a return URL. Could this be the source of my problem, at least from the Android side of things?

If there a way to have the domains for cleartext be able to reference the subnet of the local network or currently connected VPN? Or have an option to define this as a variable in the app for added security purposes/testing purposes?

Thank you for all of your hard work!

**Comment by ouchadam** [comment]
after a bit more research it turns out android only added support for mDNS lookups in android 12+ (it also requires OEMs to include support)

https://source.android.com/devices/architecture/modular-system/dns-resolver#mdns-local-resolution

| ANDROID 10 LOCAL | ANDROID 10 IP | ANDROID 12 LOCAL | 
| --- | --- | --- |
![Screenshot_20220809_180353](https://user-images.githubusercontent.com/1848238/183715175-9327b099-ea6f-4563-9684-a14848dd1778.png)|![Screenshot_20220809_180413](https://user-images.githubusercontent.com/1848238/183715172-32cae5a4-c20f-4927-b3f0-9350b27705a1.png)|![Screenshot_20220809_181342](https://user-images.githubusercontent.com/1848238/183715168-212c450c-9bb9-49b8-bfd3-c2d097b8beb4.png)

one thing to note is that the non https url must include the `http://` schema, we assume `https` by default if not provided

Using Element 1.4.31 on a Samsung fold 3 running Android 12

| ENTRY | SELECTED |
| --- | --- | 
![Screenshot_20220809_181834](https://user-images.githubusercontent.com/1848238/183715972-491c982e-9b41-47a2-be2e-c3b348ea7ca6.png)|![Screenshot_20220809_181841](https://user-images.githubusercontent.com/1848238/183715966-86f7db1a-fb22-4d8f-b070-2e1a71442886.png)|

**Comment by Burchmeister** [comment]
Awesome, so if I am able to reach a .local domain, I should be in good shape. This also correlated with the code you shared earlier. I will give this a shot as soon as I can, thank you again for your input. I will get back to you once I find out how it turned out.

**Comment by ouchadam** [comment]
yep! if you're able to access the static homeserver landing page via the browser on device then you'll be able to sign up/in with the app

I suspect there may be some dns magic you could do to get `.lan` to work but it might be simpler to change the TLD instead🤞

**Comment by Burchmeister** [comment]
![IMG_9762](https://user-images.githubusercontent.com/7854949/183785887-dfb85d55-734a-4903-985d-a0ed9b6c00f9.jpeg)
![IMG_9763](https://user-images.githubusercontent.com/7854949/183785889-9c4d2554-7f20-4166-8d01-32645b0069f4.jpeg)

So here are a few images to help this make more sense. There is a verifiable presence of matrix at the same address that is being connected to locally. The second image shows the http address being blocked on an attempted connection. If we were going to attempt to host this via a VPN, the .local suffix, correct?

**Comment by ouchadam** [comment]
thanks for the screenshots, the error in this case is expected as the ip address isn't covered by the [clear text allowance rules](https://github.com/vector-im/element-android/blob/develop/vector/src/main/res/xml/network_security_config.xml#L8) but it is good to see android is able to communicate with the server!

a very hacky workaround would be to assign the server/vpn the ip of `10.0.2.2` to match the rule for the static android emulator ip, otherwise the server will need to be hosted on a domain which is covered by the clear text rules and android is able to  discover through the dns (confirmed by accessing the homeserver landing page via the browser)

**Comment by Burchmeister** [comment]
I propose consideration of changing:

`<base-config cleartextTrafficPermitted="false" />`

 to

`<base-config cleartextTrafficPermitted="true" />`

as this would alleviate this issue, allowing testing to be more straightforward, and also open availability for use in a LOCAL/non-public-facing synapse server. This will also bring functionality on par with iOS devices, which allow this behavior to occur by default, making local instances and testing much more straightforward. 

I have read about many examples of those who use their own encryption via VPN or secondary means. Perhaps we could explore offloading this responsibility to the server managers to decide what they would ilke to use.  If the instance is private or encrypted via a strong VPN, the SSL certificate is applicable but more redundant. For local-only instances, I think this would be worth considering to potentially change. an echo prompt could even be set to allow clarification of the importance of SSL certificates for public-facing and internet based traffic. However, in these instances, SSL certs are already generally in place as they are  mandatory for security this day in age. 

Thank you immensely for your time and consideration. I appreciate all of your hard work.

**Comment by ouchadam** [comment]
will raise your suggestion in the next team session :+1:

**Comment by Heroatom** [comment]
Same issue here.
„No Network“ with all Android-based devices.
If i try to use webapp with android-based devices is the same error in all browsers.

With IOS-devices and Linux-devices it works fine.

**Comment by tyleraharrison** [comment]
Having the same issue as well

**Comment by ForteDexe** [comment]
Still have the same issue in 2024 :(

**Comment by emilyastranova** [comment]
I'm not so sure this is an issue caused by self-signed certificates. I have a LetsEncrypt certificate on my instance but still get this error unless I make the DNS record public. If I exclusively use LAN and my local DNS resolver, it's as if it completely ignores it.
